contracts/strategies/StrategyTUSDypool.sol

```diff
@@ -0,0 +1,365 @@
+        }
+    }
+
+    // Controller only function for creating additional rewards from dust
+    function withdraw(IERC20 _asset) external returns (uint256 balance) {
+        require(msg.sender == controller, "!controller");
+        require(want != address(_asset), "want");
+        require(ycrv != address(_asset), "ycrv");
+        require(yycrv != address(_asset), "yycrv");
+        balance = _asset.balanceOf(address(this));
+        _asset.safeTransfer(controller, balance);
+    }
+
+    // Withdraw partial funds, normally used with a vault withdrawal
+    function withdraw(uint256 _amount) external {
+        require(msg.sender == controller, "!controller");
+
+        rebalance();
+        uint256 _balance = IERC20(want).balanceOf(address(this));
+        if (_balance < _amount) {
+            _amount = _withdrawSome(_amount.sub(_balance));
+            _amount = _amount.add(_balance);
+            tank = 0;
+        } else {
+            if (tank >= _amount) tank = tank.sub(_amount);
+            else tank = 0;
+        }
+
+        address _vault = IController(controller).vaults(address(want));
+        require(_vault != address(0), "!vault"); // additional protection so we don't burn the funds
+        uint256 _fee = _amount.mul(withdrawalFee).div(DENOMINATOR);
+        IERC20(want).safeTransfer(IController(controller).rewards(), _fee);
+        IERC20(want).safeTransfer(_vault, _amount.sub(_fee));
+    }
+
+    function _withdrawSome(uint256 _amount) internal returns (uint256) {
+        uint256 _amnt = _amount.mul(1e18).div(ICurveFi(ypool).get_virtual_price());
+        uint256 _amt = _amnt.mul(1e18).div(yvERC20(yycrv).getPricePerFullShare());
+        uint256 _before = IERC20(ycrv).balanceOf(address(this));
+        yvERC20(yycrv).withdraw(_amt);
+        uint256 _after = IERC20(ycrv).balanceOf(address(this));
+        return _withdrawOne(_after.sub(_before));
+    }
+
+    function _withdrawOne(uint256 _amnt) internal returns (uint256) {
+        IERC20(ycrv).safeApprove(ypool, 0);
+        IERC20(ycrv).safeApprove(ypool, _amnt);
+
+        uint256 _aux = _amnt.mul(DENOMINATOR.sub(slip)).div(DENOMINATOR);
+        uint256 _t = IERC20(ycrv).totalSupply();
+        ICurveFi(ypool).remove_liquidity(
+            _amnt,
+            [
+                ICurveFi(ypool).balances(0).mul(_aux).div(_t),
+                ICurveFi(ypool).balances(1).mul(_aux).div(_t),
+                ICurveFi(ypool).balances(2).mul(_aux).div(_t),
+                ICurveFi(ypool).balances(3).mul(_aux).div(_t)
+            ]
+        );
+
+        uint256 _ydai = IERC20(ydai).balanceOf(address(this));
+        uint256 _yusdc = IERC20(yusdc).balanceOf(address(this));
+        uint256 _yusdt = IERC20(yusdt).balanceOf(address(this));
+
+        uint256 tmp;
+        if (_ydai > 0) {
+            IERC20(ydai).safeApprove(ypool, 0);
+            IERC20(ydai).safeApprove(ypool, _ydai);
+            tmp = ICurveFi(ypool).get_dy(0, 3, _ydai);
```

Comment:
To get context on curve, is the `get_dy` is like a `quote` function of the trade amount you will get?

Reply:
https://github.com/curvefi/curve-contract/blob/master/integrations.md#getting-price-for-c-tokens-cdai-cusdc

Reply:
It's for wrapped coins exchange, and this one uses D to calculate so there's no manipulation. I spent some time going all over curve codes to fully understand underlying math and implementation. This one is safe.
(Side note: Using Vyper to do deep math implementation is much more friendly than using Solidity...